EIPS/eip-3508.md

```diff
@@ -0,0 +1,100 @@
+---
+eip: 3508
+title: Transaction Data Opcodes
+author: Alex Papageorgiou (@alex-ppg)
+discussions-to: https://ethereum-magicians.org/t/eip-draft-transaction-data-opcodes/6017
+status: Draft
+type: Standards Track
+category: Core
+created: 2021-04-16
+---
+
+## Simple Summary
+
+Provide access to original transaction data.
+
+## Abstract
+
+This EIP introduces the following three EVM instructions: `ORIGINDATALOAD`, `ORIGINDATASIZE`, and `ORIGINDATACOPY`.
+
+These three instructions are meant to provide access to the original transaction's `data` payload, enabling a gas-efficient way of accessing large data payloads in cross-contract calls.
+
+## Motivation
+
+As the Ethereum development scene matures, more ambitious and complex features are introduced into smart contracts more often than not requiring the utilization of complex and at times large data structures. Given the inherent limits of the EVM, however, transporting large data structures in between contracts is a costly task that can at times lead to even futile scenarios whereby the gas consumption of such an operation is impossible to execute within the gas limit bounds as well as without sacrificing a large chunk of ETH to facilitate its gas cost.
+
+The purpose of this EIP is to render these features viable by introducing a way via which multi-contract systems are able to access the same in-memory data source without necessarily transmitting the full payload between them.
+
+This EIP enables elaborate smart contract features to become part of a larger call-chain by efficiently reading data from the original transaction payload rather than requiring the data to be passed in as call-level data. Its inclusion will mainly benefit advanced trustless schemes to manifest, such as efficient verification of Merkle Patricia trees validating the storage value of a particular Ethereum block or EVM-based layer 2 solutions.
+
+A side-effect of this change is that smart contract systems relying entirely on origin data inherently guarantee that the data they receive has not been malformed by an intermediate smart contract call.
+
+## Specification
+
+### ORIGINDATALOAD (`0x47`), ORIGINDATASIZE (`0x48`) and ORIGINDATACOPY (`0x49`)
+
+These instructions are meant to operate identically to their call-prefixed counterparts with the exception that they instead operate on the original `data` of a transaction instead of the current call's data. In detail:
+
+- ORIGINDATALOAD (`0x47`) performs identically to CALLDATALOAD (`0x35`)
+- ORIGINDATASIZE (`0x48`) performs identically to CALLDATASIZE (`0x36`)
+- ORIGINDATACOPY (`0x49`) performs identically to CALLDATACOPY (`0x37`)
```

Comment:
Change `identically` to `similarly` since these aren't strictly identical opcodes.